src/dafne/utils/sam_mask_refine.py

```diff
@@ -146,93 +200,127 @@ def generate_points_from_mask(mask):
     return np.array(point_list)
 
 
+def determine_device():
+    if GlobalConfig['USE_GPU_FOR'] != 'Autosegmentation' and torch.cuda.is_available():
+        print('SAM loaded on GPU')
+        return 'cuda'
+    print('SAM loaded on CPU')
+    return 'cpu'
+
+
 def enhance_mask(img, mask, progress_callback: Optional[Callable[[int, int], None]] = None):
-    global predictor, old_img
+    global old_img, image_embedding, predictor
 
     # if there is no mask, return the original mask
     if not mask.any():
         return mask
 
-    if GlobalConfig['SAM_MODEL'] == 'Large':
-        model_type = 'vit_h'
-    elif GlobalConfig['SAM_MODEL'] == 'Medium':
-        model_type = 'vit_l'
-    elif GlobalConfig['SAM_MODEL'] == 'Small':
-        model_type = 'vit_b'
+    device = determine_device()
+    model_choice = GlobalConfig['SAM_MODEL']
 
     def show_progress(current, maximum):
         if progress_callback is not None:
             progress_callback(current, maximum)
 
     show_progress(0, 100)
 
-    if predictor is None:
-        print('Loading SAM...')
-        predictor = load_sam(model_type, progress_callback)
+    print('Loading SAM...')
+    sam = load_sam(model_choice, progress_callback)
 
     show_progress(30, 100)
 
     if img is not old_img:
         print('Loading image...')
         old_img = img
         img_norm = img * 255 / img.max()
-        predictor.set_image(np.stack([img_norm, img_norm, img_norm], 2).astype(np.uint8))
+        image_embedding = None
+        predictor = None
 
-    show_progress(80, 100)
-
-    point_list = generate_points_from_mask(mask)
     bbox = enlarge_bounding_box(mask, GlobalConfig['SAM_BBOX_EXPAND_FACTOR'])
 
-    # generate negative points
-    negative_mask = np.logical_not(mask)
-    bbox_region = np.zeros_like(mask)
-    bbox_region[bbox[1]:bbox[3], bbox[0]:bbox[2]] = 1
-    negative_mask = np.logical_and(negative_mask, bbox_region)
-
-    negative_point_list = generate_points_from_mask(negative_mask)
-
-    t = time.perf_counter()
+    if model_choice in ['Med Sam']:
+        H, W = img.shape[:2]
+        if image_embedding is None:
+            img_3c = np.repeat(img_norm[:, :, None], 3, axis=-1) if len(img_norm.shape) == 2 else img_norm
+            img_1024 = transform.resize(img_3c, (1024, 1024), order=3, preserve_range=True, anti_aliasing=True).astype(np.uint8)
+            img_1024 = (img_1024 - img_1024.min()) / np.clip(
+                img_1024.max() - img_1024.min(), a_min=1e-8, a_max=None
+            )  # normalize to [0, 1], (H, W, 3)
+            plt.figure()
+            plt.imshow(img_1024[:, :, 0])
+            plt.pause(0.1)
```

Comment:
Also how do you run this in dafne? is there a devlopment mode to turn it on?

Reply:
Lol no, it just breaks the execution, it was a debug statement that I forgot to remove. I'm working a bit on this, to see if I manage to improve the performance, please give me some time, I will push some modification soon.

Reply:
ah gotcha, it did break the execution for me but did't show me a plot so was confused. You can take your time there's no hurry.